torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -3846,6 +3846,51 @@ def test_ddp_grad_div_uneven_inputs(self):
                     net.zero_grad()
                     torch.cuda.synchronize(device=self.rank)
 
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        @skip_if_lt_x_gpu(2)
+        def test_ddp_profiling(self):
+            batch = 3
+            dim = 10
+            num_iters = 6
+            torch.cuda.set_device(self.rank)
+            model = nn.Linear(dim, dim, bias=False)
+            inp = torch.rand(batch, dim, device=self.rank)
+            net = torch.nn.parallel.DistributedDataParallel(
+                model.cuda(self.rank), device_ids=[self.rank]
+            )
+            with torch.autograd.profiler.profile() as prof:
+                for i in range(num_iters):
+                    # Enable profiling on even iterations
+                    loss = net(inp).sum()
+                    loss.backward()
+
+            all_reduce_event_name = f"{dist.get_backend()}:all_reduce"
+            events = get_profiling_event(all_reduce_event_name, prof)
+            event_count = sum(e.count for e in events)
+            self.assertEqual(event_count, num_iters)
+            for event in events:
+                self.assertEqual(event.name, all_reduce_event_name)
+
+            # Run DDP with profiling for a few iterations, then enable profiling
+            # for a single pass, and ensure it is recorded. This tests that the
+            # thread local state is correctly updated.
+            net = torch.nn.parallel.DistributedDataParallel(
+                model.cuda(self.rank), device_ids=[self.rank]
+            )
+            for i in range(3):
+                loss = net(inp).sum()
+                loss.backward()
+            # Now enable the profiler.
+            with torch.autograd.profiler.profile() as prof:
+                loss = net(inp).sum()
+                loss.backward()
+
+            events = get_profiling_event(all_reduce_event_name, prof)
+            self.assertEqual(1, len(events))
+            self.assertEqual(1, events[0].count)
+            self.assertEqual(events[0].name, all_reduce_event_name)
```

Comment:
could we paste some profiling numbers in test plans? e.g., paste all reduce time for resnet and roberta models, would like to check whether the number is aligned with the benchmarking time measurements. thanks!

Reply:
Sure, will update the PR with measurements. Although, when comparing to benchmark, doesn't benchmark only report overall fwd/bwd/optimizer? This will provide more granular breakdowns (i.e. actual tensor ops and all_reduce), not sure if the two will be directly comparable?

Reply:
Also, note that we don't expect the events on GPU with ProcessGroupNCCL to be correct, since that work is blocked on removing cudaEvent initialization, which causes deadlock in profiler when run with DDP currently: https://github.com/pytorch/pytorch/issues/48987 (cc @ilia-cher)

Reply:
wanted to compare whether the autograd profiler all reduce time is aligned with what #51386 got:). 

I see, if the GPU measurement is still pending on fix, could you please document somewhere in our codes or so? so that users will know the GPU measure is not ready yet.

Reply:
Sounds good, let me patch that diff and compare them. I only expect it to work for Gloo right now, will add a comment for NCCL as you suggested.

Reply:
Filed issue for nccl allreduce not being profiled correctly - https://github.com/pytorch/pytorch/issues/52246